### PR TITLE
[openhabcloud] Enhanced Notification Actions

### PR DESCRIPTION
**Pull request by digitaldan**
See #16934

### CODE REVIEW

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java
Comment (maniac103) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java:75]: I guess you don't want to make this a `List<String>` to outline the number of actions being limited? I think I'd prefer doing that if there's no strong reason against it, and print some warning message if the list size exceeds 3.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java
Comment (maniac103) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java:748]: The action part is guaranteed/required to not contain any `=`? (I am thinking among the lines of `parts` potentially having more than 2 elements). It may be safer to use `indexOf('=')` and `substring()`.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java
Comment (florian-h05) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java:75]: What IMO speaks against this, is that constructing a `List<String>` is more complicated for script languages than simply passing three string args.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/BaseNotificationActionHandler.java
Comment (florian-h05) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/BaseNotificationActionHandler.java:69]: Probably https://github.com/openhab/openhab-core/blob/cb65e414453fa6d67634e0db4247bbe9d6071fac/bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java could be used here?

This would simplify the code to a few one liners:

```suggestion
        this.onClickAction = ConfigParser.valueOrElse(module.getConfiguration().get(PARAM_ON_CLICK_ACTION), String.class, null);
```

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java
Comment (maniac103) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java:75]: That applies to script languages with helper libraries, but not to rules DSL users. I'm not sure about how many there are, but forum questions suggest it's still a sizable portion of our user base?

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java
Comment (florian-h05) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java:55]: Looks better in the UI:

```suggestion
            "send a notification with icon, severity, on click, media attachment, action buttons",
```

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java
Comment (florian-h05) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java:73]: Looks better in the UI:

```suggestion
            "broadcast a notification with with icon, severity, on click, media attachment, action buttons",
```

**Review by florian-h05 — COMMENTED** [review]
Overall LGTM, UI renders the new actions correctly.

Just some minor comments:

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java:74]: Your absolutely right, i was confusing changes in the cloud code.  Jumping between the binding, IOS, Cloud, and the WebUI has my head spinning a bit.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java:748]: hmm , maybe I'm not following,  but  passing a limit of 2 ensures it only splits into 2 parts, matching on the first occurrence of '=',  so it would ignore anything in the action part.... or am i misunderstanding your feedback ?

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/NotificationAction.java:75]: After looking around the code, my assumption was that only simple primitive types would work across rules and actions ( I can't remember where in the core, but i stumbled across a class that defined these) , hence why not using more complex structures.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/BaseNotificationActionHandler.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/BaseNotificationActionHandler.java:69]: Nice, i'll look at cleaning that up, thanks for the pointer.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java:55]: agreed

**Comment by digitaldan** [comment]
Review changes (round 1)  are done, thanks!

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java
Comment (florian-h05) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java:748]: > passing a limit of 2 ensures it only splits into 2 parts, matching on the first occurrence of '='

I haven't noticed the two, then I'am fine with it :+1:

**Review by florian-h05 — APPROVED** [review]
LGTM now, great work!

**Comment by digitaldan** [comment]
Hold off on merging this, i have one more thought i'll ask on the main ticket right now on https://github.com/openhab/openhab-addons/issues/16934

**Comment by florian-h05** [comment]
I can't merge anyway in addons, so no problem ;-)

**Comment by lsiepel** [comment]
@digitaldan ping me when you are ready for a merge / review

**Comment by florian-h05** [comment]
I can't merge anyway in addons, so no problem ;-)

@digitaldan I will take care of the docs.

**Comment by florian-h05** [comment]
See #16941 for docs.
I will update my PR once this one here is merged.

**Comment by digitaldan** [comment]
> @digitaldan I will take care of the docs.

Thank You !    That a huge help!

**Comment by digitaldan** [comment]
Ok Title is in, tested and working.  Ready for review.

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java
Comment (florian-h05) [inline bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java:56]:
```suggestion
            "Sends a notification to a specific cloud user. Optionally add an icon, severity, title, on click action, media to attach, and up to 3 action buttons with a format of \"Title=Action\".",
```

**Review by florian-h05 — APPROVED** [review]
Still LGTM to me, just one minor typo:

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java
Comment (digitaldan) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/actions/NotificationActionTypeProvider.java]: Thanks!

**Comment by digitaldan** [comment]
@lsiepel this is ready !

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Comment by florian-h05** [comment]
Documentation PR #16941 is ready for merging :+1:

File: bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java
Comment (maniac103) [reply bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java:754]: I hadn't noticed the 2 either, should have paid more attention.

**Pull request merged**
